proposed/2022/NuGetExeWarnWhenSDKProjectsGetPacked.md

```diff
@@ -0,0 +1,97 @@
+- Add a dependency group for net6.0 to the nuspec
+Successfully created package 'C:\Code\Temp\SDK\SDK.1.0.0.nupkg'.
+```
+
+The command packages the single assembly, but does not generate the correct dependencies or any other metadata in the same way that the pack SDK would.
+
+Scenario 2: Multiple target framework SDK based projects, with .NET SDK installed on the machine.
+
+```console
+Attempting to build package from 'SDK.csproj'.
+MSBuild auto-detection: using msbuild version '17.4.0.37601' from 'C:\Program Files\Microsoft Visual Studio\2022\Preview\MSBuild\Current\Bin\amd64'.
+Error NU5012: Unable to find 'bin\Debug\SDK\bin\Debug\'. Make sure the project has been built.
+```
+
+NuGet.exe does not understand the concept of multitargetting and just merges the output paths.
+
+Scenario 3: SDK-based project, without the .NET SDK installed
+
+```console
+The SDK 'Microsoft.NET.Sdk' specified could not be found.  C:\Code\Temp\SDK\SDK.csproj
+```
+
+The parsing fails early.
+
+### Proposal
+
+When someone attempts to run pack on an SDK-based project, NuGet will fail with the following error:
+
+```console
+Error NU5049: NuGet.exe does not support packing SDK-based projects, use `dotnet pack` to pack this project instead.
+```
+
+### Technical explanation
+
+As described above, whenever there's no SDK installed, there's simply no accurate way for NuGet to detect that a project is SDK-based.
+We can do some XML parsing, but it is likely to be error prone. `NuGet.exe pack` is not the only command that will fail, so the customer will likely figure out that they need to install the .NET SDK. After that the same scenarios with the installed .NET SDK will follow.
+
+To determine whether a project is `.NET SDK` based, we will evaluate the project (part of the normal pack operation), and check for the `UsingMicrosoftNETSDK` property being set. If the property is set, then the project is presumed to be .NET SDK based and will be blocked from packing.
+
+## Drawbacks
+
+- The proposal here would be `breaking` for customers, but it's more than likely that the scenario for which they used nuget.exe pack on an SDK based project never worked correctly.
+
+## Rationale and alternatives
+
+- Do nothing. This would mean just acknowledging that SDK-based projects support in NuGet.exe would not be added and not do anything to help migrate users to `dotnet.exe pack` and `msbuild /t:pack`.
+- Add full pack SDK support into NuGet.exe. While not impossible, this creates maintainability considerations. dotnet.exe has a more involved support for the SDK-based PackageReference projects, and as such is considered to be the `right tool for the job`.
+- Add a warning and proceed packing.
+  - A behavior change to an error in a minor version can be considered disruptive. This is one of those special cases where the disruption might be warranted.

```

Comment:
How many customers use nuget.exe to pack SDK projects? There's a surprising amount of upvotes ([1](https://github.com/NuGet/Home/issues/4491#issuecomment-527152856), [2](https://github.com/NuGet/Home/issues/7778#issue-408582969)) for this scenario, this may be a non-trivial disruption.

Reply:
I'd venture a guess that the upvotes in both issues are from people that tried it and realized it doesn't really work properly and thus are not using anymore. 

It's difficult to know how many people are actually using `nuget.exe pack` for SDK projects successfully, since the package created does not really declare the dependencies (that's at least 1 thing that doesn't work).

The primary audience would be customers that try to pack an SDK based project with nuget.exe for the first time.

Reply:
Agreed, but I'm concerned by the potential impact of making this an error. An error is reasonable only if we know the disruption is minimal. If there are many folks using nuget.exe to pack SDK based projects, would we need to revert this error? How can we mitigate that risk? Perhaps it'd be better to play it safe and just make this a warning for now? We can always promote this to an error later once we feel more confident with the imapct.

Reply:
You summarized the concerns well, it's something hoping to get a good decision about through more feedback.

> An error is reasonable only if we know the disruption is minimal. 

That's fair. 
Another scenario where disruption may be warranted is if the scenario is irreparably broken, which this command may qualify for. 

By default, nuget.exe pack on an SDK project raises a bunch of warnings. Going through said warnings would have likely led people to figure out that nuget.exe pack is not really working the way they need it to.